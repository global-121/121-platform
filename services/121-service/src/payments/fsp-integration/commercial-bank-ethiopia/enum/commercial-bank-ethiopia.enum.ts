export enum CommercialBankEthiopiaSoapElements {
  CreditTransfer = 'CreditTransfer',
  TransactionStatus = 'TransactionStatus',
  AccountEnquiry = 'AccountEnquiry',
}
