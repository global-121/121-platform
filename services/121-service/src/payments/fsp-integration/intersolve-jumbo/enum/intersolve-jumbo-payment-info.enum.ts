export enum IntersolveJumboPaymentInfoEnum {
  firstName = 'firstName',
  lastName = 'lastName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
}
