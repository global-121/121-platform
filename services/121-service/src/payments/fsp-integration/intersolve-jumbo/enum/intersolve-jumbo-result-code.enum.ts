export enum IntersolveJumboResultCode {
  Ok = 'OK',
  ProductOwnerNotValid = 'PRODUCTOWNERNOTVALID',
  NoCustomerType = 'NOCUSTOMERTYPE',
  CustomerNotFound = 'CUSTOMERNOTFOUND',
  OrderNotValid = 'ORDERNOTVALID',
  OrderDataNotValid = 'ORDERDATANOTVALID',
  NoOrderLines = 'NOORDERLINES',
  InvalidOrderLine = 'INVALIDORDERLINE',
  NoTransactionInfo = 'NOTRANSACTIONINFO',
  ServiceFailed = 'SERVICEFAILED',
}
