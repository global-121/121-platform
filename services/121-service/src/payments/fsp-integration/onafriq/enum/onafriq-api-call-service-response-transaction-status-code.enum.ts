export enum OnafriqApiCallServiceResponseTransactionStatusCode {
  accepted = '100',
  rejected = '101',
}
