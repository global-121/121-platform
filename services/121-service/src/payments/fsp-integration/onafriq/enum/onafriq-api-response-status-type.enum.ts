export enum OnafriqApiResponseStatusType {
  success = 'success',
  duplicateThirdPartyTransIdError = 'duplicateThirdPartyTransIdError',
  genericError = 'genericError',
}
