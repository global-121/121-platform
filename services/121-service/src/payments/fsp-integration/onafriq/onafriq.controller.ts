import { Controller, HttpStatus, Post, UseGuards } from '@nestjs/common';
import { ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';

import { AuthenticatedUser } from '@121-service/src/guards/authenticated-user.decorator';
import { AuthenticatedUserGuard } from '@121-service/src/guards/authenticated-user.guard';
import { WebhookSubscribeResponseOnafriqApiDto } from '@121-service/src/payments/fsp-integration/onafriq/dtos/onafriq-api/webhook-subscribe-response-onafriq-api.dto';
import { OnafriqService } from '@121-service/src/payments/fsp-integration/onafriq/onafriq.service';

@UseGuards(AuthenticatedUserGuard)
@ApiTags('financial-service-providers/onafriq')
@Controller()
export class OnafriqController {
  public constructor(private onafriqService: OnafriqService) {}

  // NOTE: This endpoint is not tested via API-tests, as it is only used one-off in production. If it breaks, it can be fixed manually.
  @AuthenticatedUser({
    isAdmin: true,
  })
  @ApiOperation({
    summary:
      'Set callback URL for Onafriq to send transaction status updates to us. The exact callback URL is set in code, not here. In MOCK mode, this endpoint does nothing.',
  })
  @ApiResponse({
    status: HttpStatus.CREATED,
    description: 'Webhook subscribed or updated successfully',
  })
  @Post('financial-service-providers/onafriq/webhook/subscribe')
  public async subscribeWebhook(): Promise<
    WebhookSubscribeResponseOnafriqApiDto | undefined
  > {
    return await this.onafriqService.subscribeWebhook();
  }
}
