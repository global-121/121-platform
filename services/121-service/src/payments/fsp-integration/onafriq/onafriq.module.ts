import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { OnafriqTransactionEntity } from '@121-service/src/payments/fsp-integration/onafriq/entities/onafriq-transaction.entity';
import { OnafriqController } from '@121-service/src/payments/fsp-integration/onafriq/onafriq.controller';
import { OnafriqApiHelperService } from '@121-service/src/payments/fsp-integration/onafriq/services/onafriq.api.helper.service';
import { OnafriqApiService } from '@121-service/src/payments/fsp-integration/onafriq/services/onafriq.api.service';
import { OnafriqService } from '@121-service/src/payments/fsp-integration/onafriq/services/onafriq.service';
import { RedisModule } from '@121-service/src/payments/redis/redis.module';
import { QueuesRegistryModule } from '@121-service/src/queues-registry/queues-registry.module';
import { CustomHttpService } from '@121-service/src/shared/services/custom-http.service';

@Module({
  imports: [
    HttpModule,
    TypeOrmModule.forFeature([OnafriqTransactionEntity]),
    RedisModule,
    QueuesRegistryModule,
  ],
  providers: [
    OnafriqService,
    OnafriqApiService,
    OnafriqApiHelperService,
    CustomHttpService,
  ],
  exports: [OnafriqService],
  controllers: [OnafriqController],
})
export class OnafriqModule {}
