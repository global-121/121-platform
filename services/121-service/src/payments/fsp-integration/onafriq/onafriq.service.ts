import { Injectable } from '@nestjs/common';

import { PaPaymentDataDto } from '@121-service/src/payments/dto/pa-payment-data.dto';
import { FinancialServiceProviderIntegrationInterface } from '@121-service/src/payments/fsp-integration/fsp-integration.interface';
import { OnafriqApiWebhookSubscribeResponseBody } from '@121-service/src/payments/fsp-integration/onafriq/dtos/onafriq-api/onafriq-api-webhook-subscribe-response-body.dto';
import { OnafriqApiResponseStatusType } from '@121-service/src/payments/fsp-integration/onafriq/enum/onafriq-api-response-status-type.enum';
import { OnafriqError } from '@121-service/src/payments/fsp-integration/onafriq/errors/onafriq.error';
import { CreateTransactionParams } from '@121-service/src/payments/fsp-integration/onafriq/interfaces/create-transaction-params.interface';
import { OnafriqApiService } from '@121-service/src/payments/fsp-integration/onafriq/services/onafriq.api.service';

@Injectable()
export class OnafriqService
  implements FinancialServiceProviderIntegrationInterface
{
  public constructor(private readonly onafriqApiService: OnafriqApiService) {}

  /**
   * Do not use! This function was previously used to send payments.
   * It has been deprecated and should not be called anymore.
   */
  public async sendPayment(
    _paymentList: PaPaymentDataDto[],
    _programId: number,
    _paymentNr: number,
  ): Promise<void> {
    throw new Error('Method should not be called anymore.');
  }

  public async subscribeWebhook(): Promise<
    OnafriqApiWebhookSubscribeResponseBody | undefined
  > {
    return await this.onafriqApiService.subscribeWebhook();
  }

  public async createTransaction({
    transferAmount,
    phoneNumber,
    firstName,
    lastName,
    thirdPartyTransId,
  }: CreateTransactionParams): Promise<void> {
    // Simulate timeout, use this to test unintended Redis job re-attempt, by restarting 121-service during this timeout
    // 1. Simulate crash before API-call
    // await new Promise((resolve) => setTimeout(resolve, 60000));

    const mappedResponse = await this.onafriqApiService.callService({
      transferAmount,
      phoneNumber,
      firstName,
      lastName,
      thirdPartyTransId,
    });

    if (mappedResponse.status !== OnafriqApiResponseStatusType.success) {
      const errorMessage = mappedResponse.errorMessage;
      throw new OnafriqError(errorMessage!, mappedResponse.status);
    }

    // 2. Simulate crash after API call
    // await new Promise((resolve) => setTimeout(resolve, 60000));
  }
}
