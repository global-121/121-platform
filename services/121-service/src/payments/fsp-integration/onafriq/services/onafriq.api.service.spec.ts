import { HttpStatus } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';

import { OnafriqApiResponseStatusType } from '@121-service/src/payments/fsp-integration/onafriq/enum/onafriq-api-response-status-type.enum';
import { OnafriqError } from '@121-service/src/payments/fsp-integration/onafriq/errors/onafriq.error';
import { OnafriqApiHelperService } from '@121-service/src/payments/fsp-integration/onafriq/services/onafriq.api.helper.service';
import { OnafriqApiService } from '@121-service/src/payments/fsp-integration/onafriq/services/onafriq.api.service';
import { CustomHttpService } from '@121-service/src/shared/services/custom-http.service';

describe('OnafriqApiService', () => {
  let onafriqApiService: OnafriqApiService;
  let customHttpService: CustomHttpService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        OnafriqApiService,
        OnafriqApiHelperService,
        {
          provide: CustomHttpService,
          useValue: {
            post: jest.fn(),
          },
        },
      ],
    }).compile();

    onafriqApiService = module.get<OnafriqApiService>(OnafriqApiService);
    customHttpService = module.get<CustomHttpService>(CustomHttpService);
  });

  describe('callService', () => {
    it('should return generic error when OnafriqcallService returns invalid response', async () => {
      const invalidResponse = undefined;
      (customHttpService.post as jest.Mock).mockResolvedValue(invalidResponse);

      const params = {
        transferAmount: 100,
        phoneNumber: '123456789',
        firstName: 'John',
        lastName: 'Doe',
        thirdPartyTransId: 'abc123',
      };

      let error: unknown;
      try {
        await onafriqApiService.callService(params);
      } catch (e) {
        error = e;
      }
      expect(error).toBeInstanceOf(OnafriqError);
      expect((error as OnafriqError).message).toMatchSnapshot();
    });

    it('should return success status on success response', async () => {
      const validResponse = {
        status: HttpStatus.OK,
        data: {
          totalTxSent: 1,
          noTxAccepted: 1,
          noTxRejected: 0,
          details: {
            transResponse: [
              {
                thirdPartyId: 'abc123',
                status: {
                  code: '100',
                  message: 'Accepted',
                },
              },
            ],
          },
          timestamp: new Date().toISOString(),
        },
      };
      (customHttpService.post as jest.Mock).mockResolvedValue(validResponse);

      const params = {
        transferAmount: 100,
        phoneNumber: '123456789',
        firstName: 'John',
        lastName: 'Doe',
        thirdPartyTransId: 'abc123',
      };

      const result = await onafriqApiService.callService(params);

      expect(result.status).toBe(OnafriqApiResponseStatusType.success);
    });
  });

  describe('subscribeWebhook', () => {
    it('should call post twice and return the second response on success', async () => {
      const firstResponse = {
        status: HttpStatus.OK,
        statusText: 'OK',
        data: { message: 'Success' },
      };
      const secondResponse = {
        status: HttpStatus.OK,
        statusText: 'OK',
        data: { message: 'Success' },
      };
      (customHttpService.post as jest.Mock)
        .mockResolvedValueOnce(firstResponse)
        .mockResolvedValueOnce(secondResponse);

      const result = await onafriqApiService.subscribeWebhook();
      expect(customHttpService.post).toHaveBeenCalledTimes(2);
      expect(result).toEqual(secondResponse);
    });

    it('should return after first post if not 200 or message not Success', async () => {
      const firstResponse = {
        status: HttpStatus.BAD_REQUEST,
        statusText: 'Bad Request',
        data: { message: 'Error' },
      };
      (customHttpService.post as jest.Mock).mockResolvedValueOnce(
        firstResponse,
      );

      const result = await onafriqApiService.subscribeWebhook();
      expect(customHttpService.post).toHaveBeenCalledTimes(1);
      expect(result).toEqual(firstResponse);
    });
  });
});
