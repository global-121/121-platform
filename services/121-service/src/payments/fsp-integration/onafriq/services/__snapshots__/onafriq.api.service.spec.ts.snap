// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnafriqApiService callService should return generic error when OnafriqcallService returns invalid response 1`] = `"Error: Invalid Onafriq API response structure. undefined"`;
